lib/charms/magma_orchestrator_interface/v0/magma_orchestrator_interface.py

```diff
@@ -238,3 +238,54 @@ def _on_relation_changed(self, event: RelationChangedEvent) -> None:
             fluentd_address=remote_app_relation_data["fluentd_address"],
             fluentd_port=int(remote_app_relation_data["fluentd_port"]),
         )
+
+
+class OrchestratorProvides(Object):
+    """Class to be instantiated by charms providing connectivity with Orchestrator."""
+
+    def __init__(self, charm: CharmBase, relationship_name: str):
+        """Init."""
+        super().__init__(charm, relationship_name)
+        self.relationship_name = relationship_name
+        self.charm = charm
+
+    def set_orchestrator_information(
+        self,
+        root_ca_certificate: str,
+        orchestrator_address: str,
+        bootstrapper_address: str,
+        fluentd_address: str,
+        orchestrator_port: int = 443,
+        bootstrapper_port: int = 443,
+        fluentd_port: int = 24224,
+    ):
+        """Sets orchestrator information in application relation data.
+
+        Args:
+            root_ca_certificate: Orchestrator Root CA Certificate
+            orchestrator_address: Orchestrator address (ex. controller.yourdomain.com)
+            bootstrapper_address: Bootstrapper address (ex. bootstrapper-controller.yourdomain.com)
+            fluentd_address: Fluentd Address (ex. fluentd.yourdomain.com)
+            orchestrator_port: Orchestrator port (Default: 443)
+            bootstrapper_port: Bootstrapper port (Default: 443)
+            fluentd_port: Fluentd port (Default: 24224)
+
+        Returns:
+            None
+        """
+        if not self.charm.unit.is_leader():
+            raise RuntimeError("Unit must be leader to set application relation data.")
```

Comment:
Are run time errors needed? can't we just return and the leader unit handles the functionality?

Reply:
Let's say a non-leader tries to set those values, the framework would raise an error at the time of inputting the data in the databag, here we're raising earlier before even trying.

Reply:
but if you use 
```suggestion
            return
```
Would the non-leader unit be able to try to set those values?
because what I see here is that we will for sure see this exception in case of multiple units. For example if I follow the design of the dummy_test charm, let's assume it has two units, one leader unit and on non-leader unit, when it handles `orchestrator_relation_joined`, both units will run the handler, one will successfully set the data, while the other will throw an exception.
Or am I misunderstanding something?

Reply:
A proper implementation of a provider charm would only have the leader unit set this data and the non-leader wouldn't. If a non-leader tries to set relation data, it's a bad design.

Reply:
The example provider charm validates that the unit is the leader before setting it.

Reply:
okay, I'm convinced :)